### PR TITLE
Refactored DCA and fixed bugs in policy

### PR DESCRIPTION
**Pull request by milanperera**

### CODE REVIEW

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java [hunk 1]
```diff
@@ -93,10 +94,10 @@ private OAuthApplicationInfo registerApplication(RegistrationProfile profile) th
         String tokenScopes[] = new String[1];
         tokenScopes[0] = tokenScope;
 
-        oAuthApplicationInfo.addParameter("tokenScope", tokenScopes);
+        oAuthApplicationInfo.addParameter("tokenScope", Arrays.toString(tokenScopes));
```
Comment (prabathabey) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:97]: Let's have a constant defined upon the string "tokenScope".

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java [hunk 2]
```diff
@@ -134,7 +130,8 @@ private OAuthApplicationInfo registerApplication(RegistrationProfile profile) th
     }
 
     public OAuthApplicationInfo createOAuthApplication(
-            String userId, String applicationName, String callbackUrl) throws APIManagementException, IdentityException {
+            String userId, String applicationName, String callbackUrl, String grantType)
```
Comment (prabathabey) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:133]: Check the possibility of using a bean to wrap all these input parameters. Ideally, you should be able to re-use "RegistrationProfile" bean to do this.

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java (hunk 1 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:97]: Added in PR[1]
1. https://github.com/wso2/carbon-device-mgt/pull/22

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java (hunk 2 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/impl/ClientRegistrationServiceImpl.java:133]: Fixed in PR[https://github.com/wso2/carbon-device-mgt/pull/22]

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationProfile.java [hunk 3]
```diff
@@ -156,4 +157,13 @@ public void setTokenScope(String tokenScope) {
...
     }
 
+    public String getGrantType() {
+        return grantType;
+    }
+
+    public void setGrantType(String grantType) {
+        this.grantType = grantType;
+    }
+
```
Comment (prabathabey) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationProfile.java:167]: Remove unnecessary white spaces.

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/impl/PolicyInformationPointImpl.java [hunk 4]
```diff
@@ -92,16 +93,16 @@ public PIPDevice getDeviceData(DeviceIdentifier deviceIdentifier) throws PolicyM
...
+        if (pipDevice.getOwnershipType() != null && !pipDevice.getOwnershipType().isEmpty()) {
+            policies = policyFilter.filterOwnershipTypeBasedPolicies(pipDevice.getOwnershipType(), policies);
         }
         if (pipDevice.getRoles() != null) {
-            policyFilter.filterRolesBasedPolicies(pipDevice.getRoles(), policies);
+            policies = policyFilter.filterRolesBasedPolicies(pipDevice.getRoles(), policies);
         }
-        if(pipDevice.getUserIds()!=null) {
-            policyFilter.filterUserBasedPolicies(pipDevice.getUserIds(), policies);
+        if(pipDevice.getUserId() != null && !pipDevice.getUserId().isEmpty()) {
```
Comment (DilanUA) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/impl/PolicyInformationPointImpl.java:104]: keep space in between if and the condition

File: components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationProfile.java (hunk 3 above)
Comment (milanperera) [inline components/oauth-extensions/dynamic-client-manager/src/main/java/org/wso2/carbon/identity/oauth/extension/RegistrationProfile.java:167]: Fixed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/impl/PolicyInformationPointImpl.java (hunk 4 above)
Comment (milanperera) [inline components/policy-mgt/org.wso2.carbon.policy.mgt.core/src/main/java/org/wso2/carbon/policy/mgt/core/impl/PolicyInformationPointImpl.java:104]: Fixed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

**Pull request merged**